{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Basic usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook shows some basic usage of the genomic-features package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import genomic_features as gf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieving Ensembl gene annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can load annotation tables using {func}`genomic_features.ensembl.annotation`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading data from 'https://bioconductorhubs.blob.core.windows.net/annotationhub/AHEnsDbs/v98/EnsDb.Hsapiens.v98.sqlite' to file '/home/jovyan/.cache/genomic-features/87408fa36573814f9212483e6e4a939d-EnsDb.Hsapiens.v98.sqlite'.\n",
      "100%|████████████████████████████████████████| 386M/386M [00:00<00:00, 243GB/s]\n",
      "SHA256 hash of downloaded file: d02484a69b97465b66121be4c4aada06490267836a12f1c725f5d29ead3e4098\n",
      "Use this value as the 'known_hash' argument of 'pooch.retrieve' to ensure that the file hasn't changed if it is downloaded again in the future.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "EnsemblDB(organism='Homo sapiens', ensembl_release='98')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb = gf.ensembl.annotation(species=\"Hsapiens\", version=\"98\")\n",
    "ensdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These tables have been created for the [`ensembldb` Bioconductor package](https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html) {cite:p}`Rainer_2019`, and are automatically downloaded and cached from the [`AnnotationHub`](https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html) resource.\n",
    "\n",
    "We can check which Ensembl versions are available for different species using the {func}`genomic_features.ensembl.list_ensdb_annotations` util."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Ensembl_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1062</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1239</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1648</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1875</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2361</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2604</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2847</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3089</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3332</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3575</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Species  Ensembl_version\n",
       "37    Mmusculus               87\n",
       "105   Mmusculus               88\n",
       "173   Mmusculus               89\n",
       "247   Mmusculus               90\n",
       "330   Mmusculus               91\n",
       "419   Mmusculus               92\n",
       "510   Mmusculus               93\n",
       "621   Mmusculus               94\n",
       "748   Mmusculus               95\n",
       "894   Mmusculus               96\n",
       "1062  Mmusculus               97\n",
       "1239  Mmusculus               98\n",
       "1429  Mmusculus               99\n",
       "1648  Mmusculus              100\n",
       "1875  Mmusculus              101\n",
       "2118  Mmusculus              102\n",
       "2361  Mmusculus              103\n",
       "2604  Mmusculus              104\n",
       "2847  Mmusculus              105\n",
       "3089  Mmusculus              106\n",
       "3332  Mmusculus              107\n",
       "3575  Mmusculus              108\n",
       "3863  Mmusculus              109"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gf.ensembl.list_ensdb_annotations(species='Mmusculus')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using `EnsemblDB` objects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The {class}`genomic_features.ensembl.EnsemblDB` is the access point to an annotation. This is an interface to a sqlite table retrieved from AnnotationHub (as shown above). Information on provenance can be accessed via the `metadata` attribute:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Db type': 'EnsDb',\n",
       " 'Type of Gene ID': 'Ensembl Gene ID',\n",
       " 'Supporting package': 'ensembldb',\n",
       " 'Db created by': 'ensembldb package from Bioconductor',\n",
       " 'script_version': '0.3.5',\n",
       " 'Creation time': 'Mon Nov 18 21:12:12 2019',\n",
       " 'ensembl_version': '98',\n",
       " 'ensembl_host': 'localhost',\n",
       " 'Organism': 'Homo sapiens',\n",
       " 'taxonomy_id': '9606',\n",
       " 'genome_build': 'GRCh38',\n",
       " 'DBSCHEMAVERSION': '2.1'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And can be queried for genomic features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "      <td>X</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]</td>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "      <td>X</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>tenomodulin [Source:HGNC Symbol;Acc:HGNC:17757]</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50958555</td>\n",
       "      <td>20</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>dolichyl-phosphate mannosyltransferase subunit...</td>\n",
       "      <td>ENSG00000000419.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169894267</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>SCY1 like pseudokinase 3 [Source:HGNC Symbol;A...</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>chromosome 1 open reading frame 112 [Source:HG...</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name    gene_biotype  gene_seq_start  gene_seq_end   \n",
       "0  ENSG00000000003    TSPAN6  protein_coding       100627108     100639991  \\\n",
       "1  ENSG00000000005      TNMD  protein_coding       100584936     100599885   \n",
       "2  ENSG00000000419      DPM1  protein_coding        50934867      50958555   \n",
       "3  ENSG00000000457     SCYL3  protein_coding       169849631     169894267   \n",
       "4  ENSG00000000460  C1orf112  protein_coding       169662007     169854080   \n",
       "\n",
       "  seq_name  seq_strand seq_coord_system   \n",
       "0        X          -1       chromosome  \\\n",
       "1        X           1       chromosome   \n",
       "2       20          -1       chromosome   \n",
       "3        1          -1       chromosome   \n",
       "4        1           1       chromosome   \n",
       "\n",
       "                                         description     gene_id_version  \n",
       "0  tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]  ENSG00000000003.15  \n",
       "1    tenomodulin [Source:HGNC Symbol;Acc:HGNC:17757]   ENSG00000000005.6  \n",
       "2  dolichyl-phosphate mannosyltransferase subunit...  ENSG00000000419.12  \n",
       "3  SCY1 like pseudokinase 3 [Source:HGNC Symbol;A...  ENSG00000000457.14  \n",
       "4  chromosome 1 open reading frame 112 [Source:HG...  ENSG00000000460.17  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ensdb.genes()\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "{mod}`genomic_features.filters` defines a number of filters to use with these annotations. You can filter by specific columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Phe (UUU/C) [Sour...</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Val (GUN) [Source...</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210100</td>\n",
       "      <td>MT-TI</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Ile (AUU/C) [Sour...</td>\n",
       "      <td>ENSG00000210100.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>MT-TQ</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>MT</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Gln (CAA/G) [Sour...</td>\n",
       "      <td>ENSG00000210107.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name gene_biotype  gene_seq_start  gene_seq_end   \n",
       "0  ENSG00000209082    MT-TL1      Mt_tRNA            3230          3304  \\\n",
       "1  ENSG00000210049     MT-TF      Mt_tRNA             577           647   \n",
       "2  ENSG00000210077     MT-TV      Mt_tRNA            1602          1670   \n",
       "3  ENSG00000210100     MT-TI      Mt_tRNA            4263          4331   \n",
       "4  ENSG00000210107     MT-TQ      Mt_tRNA            4329          4400   \n",
       "\n",
       "  seq_name  seq_strand seq_coord_system   \n",
       "0       MT           1       chromosome  \\\n",
       "1       MT           1       chromosome   \n",
       "2       MT           1       chromosome   \n",
       "3       MT           1       chromosome   \n",
       "4       MT          -1       chromosome   \n",
       "\n",
       "                                         description    gene_id_version  \n",
       "0  mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...  ENSG00000209082.1  \n",
       "1  mitochondrially encoded tRNA-Phe (UUU/C) [Sour...  ENSG00000210049.1  \n",
       "2  mitochondrially encoded tRNA-Val (GUN) [Source...  ENSG00000210077.1  \n",
       "3  mitochondrially encoded tRNA-Ile (AUU/C) [Sour...  ENSG00000210100.1  \n",
       "4  mitochondrially encoded tRNA-Gln (CAA/G) [Sour...  ENSG00000210107.1  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(filter=gf.filters.GeneBioTypeFilter(\"Mt_tRNA\")).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or by genomic range:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1 [Source:HGNC Sym...</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>WASP family homolog 7, pseudogene [Source:HGNC...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>microRNA 6859-1 [Source:HGNC Symbol;Acc:HGNC:5...</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id  gene_name                        gene_biotype   \n",
       "0  ENSG00000223972    DDX11L1  transcribed_unprocessed_pseudogene  \\\n",
       "1  ENSG00000227232     WASH7P              unprocessed_pseudogene   \n",
       "2  ENSG00000278267  MIR6859-1                               miRNA   \n",
       "\n",
       "   gene_seq_start  gene_seq_end seq_name  seq_strand seq_coord_system   \n",
       "0           11869         14409        1           1       chromosome  \\\n",
       "1           14404         29570        1          -1       chromosome   \n",
       "2           17369         17436        1          -1       chromosome   \n",
       "\n",
       "                                         description    gene_id_version  \n",
       "0  DEAD/H-box helicase 11 like 1 [Source:HGNC Sym...  ENSG00000223972.5  \n",
       "1  WASP family homolog 7, pseudogene [Source:HGNC...  ENSG00000227232.5  \n",
       "2  microRNA 6859-1 [Source:HGNC Symbol;Acc:HGNC:5...  ENSG00000278267.1  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(filter=gf.filters.GeneRangesFilter(\"1:10000-20000\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Logical operations (`&`, `|`, and `~`) on filters are also possible:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_id, gene_name, gene_biotype, gene_seq_start, gene_seq_end, seq_name, seq_strand, seq_coord_system, description, gene_id_version]\n",
       "Index: []"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(\n",
    "    filter=gf.filters.GeneBioTypeFilter(\"lncRNA\") & gf.filters.GeneRangesFilter(\"1:10000-20000\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Column selectors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the `cols` argument, you can get annotations from other tables in the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>O43657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000612152</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSP00000482130</td>\n",
       "      <td>A0A087WYV6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000614008</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSP00000482894</td>\n",
       "      <td>A0A087WZU5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENST00000373031</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "      <td>Q9H2S6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENST00000371588</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>ENSP00000360644</td>\n",
       "      <td>A0A0S2Z4Y5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id            tx_id gene_name       protein_id  uniprot_id\n",
       "0  ENSG00000000003  ENST00000373020    TSPAN6  ENSP00000362111      O43657\n",
       "1  ENSG00000000003  ENST00000612152    TSPAN6  ENSP00000482130  A0A087WYV6\n",
       "2  ENSG00000000003  ENST00000614008    TSPAN6  ENSP00000482894  A0A087WZU5\n",
       "3  ENSG00000000005  ENST00000373031      TNMD  ENSP00000362122      Q9H2S6\n",
       "4  ENSG00000000419  ENST00000371588      DPM1  ENSP00000360644  A0A0S2Z4Y5"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(cols=[\"gene_id\", \"tx_id\", \"gene_name\", \"protein_id\", \"uniprot_id\"]).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `chromosomes()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Information on chromosome length for this annotation (useful for downstream operations) is also available through the `chromosomes` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>is_circular</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X</td>\n",
       "      <td>156040895</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20</td>\n",
       "      <td>64444167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>170805979</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>198295559</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>LRG_311</td>\n",
       "      <td>115492</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>LRG_721</td>\n",
       "      <td>33396</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>LRG_741</td>\n",
       "      <td>231167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>LRG_763</td>\n",
       "      <td>176286</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>LRG_93</td>\n",
       "      <td>22459</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>454 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seq_name  seq_length  is_circular\n",
       "0          X   156040895            0\n",
       "1         20    64444167            0\n",
       "2          1   248956422            0\n",
       "3          6   170805979            0\n",
       "4          3   198295559            0\n",
       "..       ...         ...          ...\n",
       "449  LRG_311      115492            0\n",
       "450  LRG_721       33396            0\n",
       "451  LRG_741      231167            0\n",
       "452  LRG_763      176286            0\n",
       "453   LRG_93       22459            0\n",
       "\n",
       "[454 rows x 3 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.chromosomes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding annotations to an AnnData object\n",
    "\n",
    "We provide a utility function `annotate` to merge a table of feature annotations from an scverse dataset and a gene annotation table. In this example we use subsampled scRNA-seq data from [Litviňuková et al. 2020](https://www.nature.com/articles/s41586-020-2797-4), where reads were mapped to the hg38 reference genome with [CellRanger v3.0.1](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/3.0), which uses the Ensembl release 98 for gene annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m File data/hca_subsampled_20k.h5ad already downloaded                                                      \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import scvi\n",
    "adata = scvi.data.heart_cell_atlas_subsampled()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var = adata.var[['gene_ids-Harvard-Nuclei']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_985/997274008.py:1: UserWarning: Missing annotations for 156/26662 vars\n",
      "  annotated_var = gf.annotate(adata.var, genes, var_on='gene_ids-Harvard-Nuclei')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids-Harvard-Nuclei</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>89295.0</td>\n",
       "      <td>133723.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC114498.1</th>\n",
       "      <td>ENSG00000235146</td>\n",
       "      <td>ENSG00000235146</td>\n",
       "      <td>AC114498.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>587629.0</td>\n",
       "      <td>594768.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>ENSG00000235146.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL669831.2</th>\n",
       "      <td>ENSG00000229905</td>\n",
       "      <td>ENSG00000229905</td>\n",
       "      <td>AL669831.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>760911.0</td>\n",
       "      <td>761989.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>ENSG00000229905.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL669831.5</th>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>ENSG00000237491</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>778747.0</td>\n",
       "      <td>810065.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>long intergenic non-protein coding RNA 1409 [S...</td>\n",
       "      <td>ENSG00000237491.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM87B</th>\n",
       "      <td>ENSG00000177757</td>\n",
       "      <td>ENSG00000177757</td>\n",
       "      <td>FAM87B</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>817371.0</td>\n",
       "      <td>819837.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>family with sequence similarity 87 member B [S...</td>\n",
       "      <td>ENSG00000177757.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>AC007325.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>138082.0</td>\n",
       "      <td>161852.0</td>\n",
       "      <td>KI270734.1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>proline dehydrogenase 1, mitochondrial [Source...</td>\n",
       "      <td>ENSG00000277196.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BX072566.1</th>\n",
       "      <td>ENSG00000277630</td>\n",
       "      <td>ENSG00000277630</td>\n",
       "      <td>BX072566.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>108007.0</td>\n",
       "      <td>139659.0</td>\n",
       "      <td>GL000213.1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>POTE ankyrin domain family member D-like [Sour...</td>\n",
       "      <td>ENSG00000277630.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL354822.1</th>\n",
       "      <td>ENSG00000278384</td>\n",
       "      <td>ENSG00000278384</td>\n",
       "      <td>AL354822.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>51867.0</td>\n",
       "      <td>54893.0</td>\n",
       "      <td>GL000218.1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>NULL</td>\n",
       "      <td>ENSG00000278384.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004556.1</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>AC004556.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2585.0</td>\n",
       "      <td>11802.0</td>\n",
       "      <td>KI270721.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>39S ribosomal protein L23, mitochondrial [Sour...</td>\n",
       "      <td>ENSG00000276345.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4612.0</td>\n",
       "      <td>29626.0</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>neuroblastoma breakpoint family member 1 [Sour...</td>\n",
       "      <td>ENSG00000271254.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26662 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_ids-Harvard-Nuclei          gene_id   gene_name   \n",
       "AL627309.1         ENSG00000238009  ENSG00000238009  AL627309.1  \\\n",
       "AC114498.1         ENSG00000235146  ENSG00000235146  AC114498.1   \n",
       "AL669831.2         ENSG00000229905  ENSG00000229905  AL669831.2   \n",
       "AL669831.5         ENSG00000237491  ENSG00000237491   LINC01409   \n",
       "FAM87B             ENSG00000177757  ENSG00000177757      FAM87B   \n",
       "...                            ...              ...         ...   \n",
       "AC007325.2         ENSG00000277196  ENSG00000277196  AC007325.2   \n",
       "BX072566.1         ENSG00000277630  ENSG00000277630  BX072566.1   \n",
       "AL354822.1         ENSG00000278384  ENSG00000278384  AL354822.1   \n",
       "AC004556.1         ENSG00000276345  ENSG00000276345  AC004556.3   \n",
       "AC240274.1         ENSG00000271254  ENSG00000271254  AC240274.1   \n",
       "\n",
       "              gene_biotype  gene_seq_start  gene_seq_end    seq_name   \n",
       "AL627309.1          lncRNA         89295.0      133723.0           1  \\\n",
       "AC114498.1          lncRNA        587629.0      594768.0           1   \n",
       "AL669831.2          lncRNA        760911.0      761989.0           1   \n",
       "AL669831.5          lncRNA        778747.0      810065.0           1   \n",
       "FAM87B              lncRNA        817371.0      819837.0           1   \n",
       "...                    ...             ...           ...         ...   \n",
       "AC007325.2  protein_coding        138082.0      161852.0  KI270734.1   \n",
       "BX072566.1  protein_coding        108007.0      139659.0  GL000213.1   \n",
       "AL354822.1  protein_coding         51867.0       54893.0  GL000218.1   \n",
       "AC004556.1  protein_coding          2585.0       11802.0  KI270721.1   \n",
       "AC240274.1  protein_coding          4612.0       29626.0  KI270711.1   \n",
       "\n",
       "            seq_strand seq_coord_system   \n",
       "AL627309.1        -1.0       chromosome  \\\n",
       "AC114498.1         1.0       chromosome   \n",
       "AL669831.2         1.0       chromosome   \n",
       "AL669831.5         1.0       chromosome   \n",
       "FAM87B             1.0       chromosome   \n",
       "...                ...              ...   \n",
       "AC007325.2        -1.0         scaffold   \n",
       "BX072566.1        -1.0         scaffold   \n",
       "AL354822.1        -1.0         scaffold   \n",
       "AC004556.1         1.0         scaffold   \n",
       "AC240274.1        -1.0         scaffold   \n",
       "\n",
       "                                                  description   \n",
       "AL627309.1                                   novel transcript  \\\n",
       "AC114498.1                                   novel transcript   \n",
       "AL669831.2                                   novel transcript   \n",
       "AL669831.5  long intergenic non-protein coding RNA 1409 [S...   \n",
       "FAM87B      family with sequence similarity 87 member B [S...   \n",
       "...                                                       ...   \n",
       "AC007325.2  proline dehydrogenase 1, mitochondrial [Source...   \n",
       "BX072566.1  POTE ankyrin domain family member D-like [Sour...   \n",
       "AL354822.1                                               NULL   \n",
       "AC004556.1  39S ribosomal protein L23, mitochondrial [Sour...   \n",
       "AC240274.1  neuroblastoma breakpoint family member 1 [Sour...   \n",
       "\n",
       "               gene_id_version  \n",
       "AL627309.1   ENSG00000238009.6  \n",
       "AC114498.1   ENSG00000235146.2  \n",
       "AL669831.2   ENSG00000229905.1  \n",
       "AL669831.5  ENSG00000237491.10  \n",
       "FAM87B       ENSG00000177757.2  \n",
       "...                        ...  \n",
       "AC007325.2   ENSG00000277196.4  \n",
       "BX072566.1   ENSG00000277630.4  \n",
       "AL354822.1   ENSG00000278384.1  \n",
       "AC004556.1   ENSG00000276345.1  \n",
       "AC240274.1   ENSG00000271254.6  \n",
       "\n",
       "[26662 rows x 11 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotated_var = gf.annotate(adata.var, genes, var_on='gene_ids-Harvard-Nuclei')\n",
    "annotated_var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var = annotated_var.copy()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
