{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Basic usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook shows some basic usage of the genomic-features package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import genomic_features as gf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieving Ensembl gene annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can load annotation tables using {func}`genomic_features.ensembl.annotation`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EnsemblDB(organism='Homo sapiens', ensembl_release='108')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb = gf.ensembl.annotation(species=\"Hsapiens\", version=\"108\")\n",
    "ensdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These tables have been created for the [`ensembldb` Bioconductor package](https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html) {cite:p}`Rainer_2019`, and are automatically downloaded and cached from the [`AnnotationHub`](https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html) resource.\n",
    "\n",
    "We can check which Ensembl versions are available for different species using the {func}`genomic_features.ensembl.list_ensdb_annotations` util."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Ensembl_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>510</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1062</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1239</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1648</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1875</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2361</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2604</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2847</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3089</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3332</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3575</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863</th>\n",
       "      <td>Mmusculus</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Species  Ensembl_version\n",
       "37    Mmusculus               87\n",
       "105   Mmusculus               88\n",
       "173   Mmusculus               89\n",
       "247   Mmusculus               90\n",
       "330   Mmusculus               91\n",
       "419   Mmusculus               92\n",
       "510   Mmusculus               93\n",
       "621   Mmusculus               94\n",
       "748   Mmusculus               95\n",
       "894   Mmusculus               96\n",
       "1062  Mmusculus               97\n",
       "1239  Mmusculus               98\n",
       "1429  Mmusculus               99\n",
       "1648  Mmusculus              100\n",
       "1875  Mmusculus              101\n",
       "2118  Mmusculus              102\n",
       "2361  Mmusculus              103\n",
       "2604  Mmusculus              104\n",
       "2847  Mmusculus              105\n",
       "3089  Mmusculus              106\n",
       "3332  Mmusculus              107\n",
       "3575  Mmusculus              108\n",
       "3863  Mmusculus              109"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gf.ensembl.list_ensdb_annotations(species=\"Mmusculus\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using `EnsemblDB` objects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The {class}`genomic_features.ensembl.EnsemblDB` is the access point to an annotation. This is an interface to a sqlite table retrieved from AnnotationHub (as shown above). Information on provenance can be accessed via the `metadata` attribute:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Db type': 'EnsDb',\n",
       " 'Type of Gene ID': 'Ensembl Gene ID',\n",
       " 'Supporting package': 'ensembldb',\n",
       " 'Db created by': 'ensembldb package from Bioconductor',\n",
       " 'script_version': '0.3.7',\n",
       " 'Creation time': 'Fri Oct 28 05:24:43 2022',\n",
       " 'ensembl_version': '108',\n",
       " 'ensembl_host': 'localhost',\n",
       " 'Organism': 'Homo sapiens',\n",
       " 'taxonomy_id': '9606',\n",
       " 'genome_build': 'GRCh38',\n",
       " 'DBSCHEMAVERSION': '2.2'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And can be queried for genomic features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "      <th>canonical_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100627108</td>\n",
       "      <td>100639991</td>\n",
       "      <td>X</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]</td>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "      <td>ENST00000373020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100584936</td>\n",
       "      <td>100599885</td>\n",
       "      <td>X</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>tenomodulin [Source:HGNC Symbol;Acc:HGNC:17757]</td>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>ENST00000373031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>50934867</td>\n",
       "      <td>50959140</td>\n",
       "      <td>20</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>dolichyl-phosphate mannosyltransferase subunit...</td>\n",
       "      <td>ENSG00000000419.14</td>\n",
       "      <td>ENST00000371588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169894267</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>SCY1 like pseudokinase 3 [Source:HGNC Symbol;A...</td>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>ENST00000367771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>169662007</td>\n",
       "      <td>169854080</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>chromosome 1 open reading frame 112 [Source:HG...</td>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>ENST00000359326</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name    gene_biotype  gene_seq_start  gene_seq_end   \n",
       "0  ENSG00000000003    TSPAN6  protein_coding       100627108     100639991  \\\n",
       "1  ENSG00000000005      TNMD  protein_coding       100584936     100599885   \n",
       "2  ENSG00000000419      DPM1  protein_coding        50934867      50959140   \n",
       "3  ENSG00000000457     SCYL3  protein_coding       169849631     169894267   \n",
       "4  ENSG00000000460  C1orf112  protein_coding       169662007     169854080   \n",
       "\n",
       "  seq_name  seq_strand seq_coord_system   \n",
       "0        X          -1       chromosome  \\\n",
       "1        X           1       chromosome   \n",
       "2       20          -1       chromosome   \n",
       "3        1          -1       chromosome   \n",
       "4        1           1       chromosome   \n",
       "\n",
       "                                         description     gene_id_version   \n",
       "0  tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]  ENSG00000000003.15  \\\n",
       "1    tenomodulin [Source:HGNC Symbol;Acc:HGNC:17757]   ENSG00000000005.6   \n",
       "2  dolichyl-phosphate mannosyltransferase subunit...  ENSG00000000419.14   \n",
       "3  SCY1 like pseudokinase 3 [Source:HGNC Symbol;A...  ENSG00000000457.14   \n",
       "4  chromosome 1 open reading frame 112 [Source:HG...  ENSG00000000460.17   \n",
       "\n",
       "  canonical_transcript  \n",
       "0      ENST00000373020  \n",
       "1      ENST00000373031  \n",
       "2      ENST00000371588  \n",
       "3      ENST00000367771  \n",
       "4      ENST00000359326  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ensdb.genes()\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "{mod}`genomic_features.filters` defines a number of filters to use with these annotations. You can filter by specific columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "      <th>canonical_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>3230</td>\n",
       "      <td>3304</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>577</td>\n",
       "      <td>647</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Phe (UUU/C) [Sour...</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>1602</td>\n",
       "      <td>1670</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Val (GUN) [Source...</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210100</td>\n",
       "      <td>MT-TI</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>4263</td>\n",
       "      <td>4331</td>\n",
       "      <td>MT</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Ile (AUU/C) [Sour...</td>\n",
       "      <td>ENSG00000210100.1</td>\n",
       "      <td>ENST00000387365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>MT-TQ</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>4329</td>\n",
       "      <td>4400</td>\n",
       "      <td>MT</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>mitochondrially encoded tRNA-Gln (CAA/G) [Sour...</td>\n",
       "      <td>ENSG00000210107.1</td>\n",
       "      <td>ENST00000387372</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name gene_biotype  gene_seq_start  gene_seq_end   \n",
       "0  ENSG00000209082    MT-TL1      Mt_tRNA            3230          3304  \\\n",
       "1  ENSG00000210049     MT-TF      Mt_tRNA             577           647   \n",
       "2  ENSG00000210077     MT-TV      Mt_tRNA            1602          1670   \n",
       "3  ENSG00000210100     MT-TI      Mt_tRNA            4263          4331   \n",
       "4  ENSG00000210107     MT-TQ      Mt_tRNA            4329          4400   \n",
       "\n",
       "  seq_name  seq_strand seq_coord_system   \n",
       "0       MT           1       chromosome  \\\n",
       "1       MT           1       chromosome   \n",
       "2       MT           1       chromosome   \n",
       "3       MT           1       chromosome   \n",
       "4       MT          -1       chromosome   \n",
       "\n",
       "                                         description    gene_id_version   \n",
       "0  mitochondrially encoded tRNA-Leu (UUA/G) 1 [So...  ENSG00000209082.1  \\\n",
       "1  mitochondrially encoded tRNA-Phe (UUU/C) [Sour...  ENSG00000210049.1   \n",
       "2  mitochondrially encoded tRNA-Val (GUN) [Source...  ENSG00000210077.1   \n",
       "3  mitochondrially encoded tRNA-Ile (AUU/C) [Sour...  ENSG00000210100.1   \n",
       "4  mitochondrially encoded tRNA-Gln (CAA/G) [Sour...  ENSG00000210107.1   \n",
       "\n",
       "  canonical_transcript  \n",
       "0      ENST00000386347  \n",
       "1      ENST00000387314  \n",
       "2      ENST00000387342  \n",
       "3      ENST00000387365  \n",
       "4      ENST00000387372  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(filter=gf.filters.GeneBioTypeFilter(\"Mt_tRNA\")).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or by genomic range:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "      <th>canonical_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1 (pseudogene) [So...</td>\n",
       "      <td>ENSG00000223972.6</td>\n",
       "      <td>ENST00000450305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>WASP family homolog 7, pseudogene [Source:HGNC...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>microRNA 6859-1 [Source:HGNC Symbol;Acc:HGNC:5...</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>DEAD/H-box helicase 11 like 2 (pseudogene) [So...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>ENST00000456328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id  gene_name                        gene_biotype   \n",
       "0  ENSG00000223972    DDX11L1  transcribed_unprocessed_pseudogene  \\\n",
       "1  ENSG00000227232     WASH7P              unprocessed_pseudogene   \n",
       "2  ENSG00000278267  MIR6859-1                               miRNA   \n",
       "3  ENSG00000290825    DDX11L2                              lncRNA   \n",
       "\n",
       "   gene_seq_start  gene_seq_end seq_name  seq_strand seq_coord_system   \n",
       "0           12010         13670        1           1       chromosome  \\\n",
       "1           14404         29570        1          -1       chromosome   \n",
       "2           17369         17436        1          -1       chromosome   \n",
       "3           11869         14409        1           1       chromosome   \n",
       "\n",
       "                                         description    gene_id_version   \n",
       "0  DEAD/H-box helicase 11 like 1 (pseudogene) [So...  ENSG00000223972.6  \\\n",
       "1  WASP family homolog 7, pseudogene [Source:HGNC...  ENSG00000227232.5   \n",
       "2  microRNA 6859-1 [Source:HGNC Symbol;Acc:HGNC:5...  ENSG00000278267.1   \n",
       "3  DEAD/H-box helicase 11 like 2 (pseudogene) [So...  ENSG00000290825.1   \n",
       "\n",
       "  canonical_transcript  \n",
       "0      ENST00000450305  \n",
       "1      ENST00000488147  \n",
       "2      ENST00000619216  \n",
       "3      ENST00000456328  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(filter=gf.filters.GeneRangesFilter(\"1:10000-20000\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Logical operations (`&`, `|`, and `~`) on filters are also possible:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "      <th>canonical_transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>DEAD/H-box helicase 11 like 2 (pseudogene) [So...</td>\n",
       "      <td>ENSG00000290825.1</td>\n",
       "      <td>ENST00000456328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name gene_biotype  gene_seq_start  gene_seq_end   \n",
       "0  ENSG00000290825   DDX11L2       lncRNA           11869         14409  \\\n",
       "\n",
       "  seq_name  seq_strand seq_coord_system   \n",
       "0        1           1       chromosome  \\\n",
       "\n",
       "                                         description    gene_id_version   \n",
       "0  DEAD/H-box helicase 11 like 2 (pseudogene) [So...  ENSG00000290825.1  \\\n",
       "\n",
       "  canonical_transcript  \n",
       "0      ENST00000456328  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(\n",
    "    filter=gf.filters.GeneBioTypeFilter(\"lncRNA\")\n",
    "    & gf.filters.GeneRangesFilter(\"1:10000-20000\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Column selectors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the `cols` argument, you can get annotations from other tables in the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000373020</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSP00000362111</td>\n",
       "      <td>O43657.176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000612152</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSP00000482130</td>\n",
       "      <td>A0A087WYV6.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>ENST00000614008</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSP00000482894</td>\n",
       "      <td>A0A087WZU5.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>ENST00000373031</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSP00000362122</td>\n",
       "      <td>Q9H2S6.148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>ENST00000466152</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>ENSP00000507119</td>\n",
       "      <td>A0A804HIK9.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id            tx_id gene_name       protein_id     uniprot_id\n",
       "0  ENSG00000000003  ENST00000373020    TSPAN6  ENSP00000362111     O43657.176\n",
       "1  ENSG00000000003  ENST00000612152    TSPAN6  ENSP00000482130  A0A087WYV6.48\n",
       "2  ENSG00000000003  ENST00000614008    TSPAN6  ENSP00000482894  A0A087WZU5.51\n",
       "3  ENSG00000000005  ENST00000373031      TNMD  ENSP00000362122     Q9H2S6.148\n",
       "4  ENSG00000000419  ENST00000466152      DPM1  ENSP00000507119   A0A804HIK9.2"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.genes(cols=[\"gene_id\", \"tx_id\", \"gene_name\", \"protein_id\", \"uniprot_id\"]).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `chromosomes()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Information on chromosome length for this annotation (useful for downstream operations) is also available through the `chromosomes` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>is_circular</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X</td>\n",
       "      <td>156040895</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20</td>\n",
       "      <td>64444167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>170805979</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>198295559</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>LRG_741</td>\n",
       "      <td>231167</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>LRG_763</td>\n",
       "      <td>176286</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>LRG_792</td>\n",
       "      <td>42144</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>LRG_793</td>\n",
       "      <td>38439</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>LRG_93</td>\n",
       "      <td>22459</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>457 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seq_name  seq_length  is_circular\n",
       "0          X   156040895            0\n",
       "1         20    64444167            0\n",
       "2          1   248956422            0\n",
       "3          6   170805979            0\n",
       "4          3   198295559            0\n",
       "..       ...         ...          ...\n",
       "452  LRG_741      231167            0\n",
       "453  LRG_763      176286            0\n",
       "454  LRG_792       42144            0\n",
       "455  LRG_793       38439            0\n",
       "456   LRG_93       22459            0\n",
       "\n",
       "[457 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensdb.chromosomes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding annotations to an AnnData object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "pbmc = sc.datasets.pbmc3k()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-10</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.8</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_ids\n",
       "index                        \n",
       "MIR1302-10    ENSG00000243485\n",
       "FAM138A       ENSG00000237613\n",
       "OR4F5         ENSG00000186092\n",
       "RP11-34P13.7  ENSG00000238009\n",
       "RP11-34P13.8  ENSG00000239945"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pbmc.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc.var = pd.merge(\n",
    "    (\n",
    "        pbmc.var.reset_index().rename(\n",
    "            columns={\"gene_ids\": \"gene_id\", \"index\": \"orig_gene_name\"}\n",
    "        )\n",
    "    ),\n",
    "    genes,\n",
    "    on=\"gene_id\",\n",
    "    how=\"left\",\n",
    ").set_index(\"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_gene_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_seq_start</th>\n",
       "      <th>gene_seq_end</th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_strand</th>\n",
       "      <th>seq_coord_system</th>\n",
       "      <th>description</th>\n",
       "      <th>gene_id_version</th>\n",
       "      <th>canonical_transcript</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-10</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>29554.0</td>\n",
       "      <td>31109.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>MIR1302-2 host gene [Source:HGNC Symbol;Acc:HG...</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>34554.0</td>\n",
       "      <td>36081.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>family with sequence similarity 138 member A [...</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>65419.0</td>\n",
       "      <td>71585.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>olfactory receptor family 4 subfamily F member...</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>ENST00000641515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>RP11-34P13.7</td>\n",
       "      <td></td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>89295.0</td>\n",
       "      <td>133723.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000477740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td></td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>89551.0</td>\n",
       "      <td>91105.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>novel transcript</td>\n",
       "      <td>ENSG00000239945.1</td>\n",
       "      <td>ENST00000495576</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                orig_gene_name    gene_name    gene_biotype  gene_seq_start   \n",
       "gene_id                                                                       \n",
       "ENSG00000243485     MIR1302-10  MIR1302-2HG          lncRNA         29554.0  \\\n",
       "ENSG00000237613        FAM138A      FAM138A          lncRNA         34554.0   \n",
       "ENSG00000186092          OR4F5        OR4F5  protein_coding         65419.0   \n",
       "ENSG00000238009   RP11-34P13.7                       lncRNA         89295.0   \n",
       "ENSG00000239945   RP11-34P13.8                       lncRNA         89551.0   \n",
       "\n",
       "                 gene_seq_end seq_name  seq_strand seq_coord_system   \n",
       "gene_id                                                               \n",
       "ENSG00000243485       31109.0        1         1.0       chromosome  \\\n",
       "ENSG00000237613       36081.0        1        -1.0       chromosome   \n",
       "ENSG00000186092       71585.0        1         1.0       chromosome   \n",
       "ENSG00000238009      133723.0        1        -1.0       chromosome   \n",
       "ENSG00000239945       91105.0        1        -1.0       chromosome   \n",
       "\n",
       "                                                       description   \n",
       "gene_id                                                              \n",
       "ENSG00000243485  MIR1302-2 host gene [Source:HGNC Symbol;Acc:HG...  \\\n",
       "ENSG00000237613  family with sequence similarity 138 member A [...   \n",
       "ENSG00000186092  olfactory receptor family 4 subfamily F member...   \n",
       "ENSG00000238009                                   novel transcript   \n",
       "ENSG00000239945                                   novel transcript   \n",
       "\n",
       "                   gene_id_version canonical_transcript  \n",
       "gene_id                                                  \n",
       "ENSG00000243485  ENSG00000243485.5      ENST00000473358  \n",
       "ENSG00000237613  ENSG00000237613.2      ENST00000417324  \n",
       "ENSG00000186092  ENSG00000186092.7      ENST00000641515  \n",
       "ENSG00000238009  ENSG00000238009.6      ENST00000477740  \n",
       "ENSG00000239945  ENSG00000239945.1      ENST00000495576  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pbmc.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:genomic-annotations-dev]",
   "language": "python",
   "name": "conda-env-genomic-annotations-dev-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "vscode": {
   "interpreter": {
    "hash": "ae6466e8d4f517858789b5c9e8f0ed238fb8964458a36305fca7bddc149e9c64"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
